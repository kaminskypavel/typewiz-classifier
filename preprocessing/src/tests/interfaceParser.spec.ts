import {
	getInterfaceName,
	getInterfaceProperties,
	getPropertyKind,
	getPropertyName,
	getPropertyQuestion,
	parseInterface
} from '../interfaceParser';
import {example1, example2, example3, example4, example5} from './fixtures/mockInterfaces';

describe('#getInterfaceName', () => {
	it('should return the interface name', () => {
		expect(getInterfaceName(example1)).toBe('IExample1');
		expect(getInterfaceName(example2)).toBe('IExample2');
		expect(getInterfaceName(example3)).toBe('IExample3');
		expect(getInterfaceName(example4)).toBe('IExample4');
		expect(getInterfaceName('interface example {}')).toBe('IExample');
		expect(getInterfaceName('interface Iexample {}')).toBe('IExample');
		expect(getInterfaceName('interface IExample {}')).toBe('IExample');
		expect(getInterfaceName('interface iExample {}')).toBe('IExample');
		expect(getInterfaceName('interface iexample {}')).toBe('IExample');
		expect(getInterfaceName('interface IReducerInterface {}')).toBe('IReducerInterface');
		expect(getInterfaceName('interface reducer_interface {}')).toBe('IReducerInterface');
	});
});

describe('#getProperties', () => {
	it('should find the correct number on nodes', () => {
		expect(getInterfaceProperties(example1).length).toEqual(0);
		expect(getInterfaceProperties(example2).length).toEqual(5);
		expect(getInterfaceProperties(example3).length).toEqual(6);
		expect(getInterfaceProperties(example4).length).toEqual(9);
	});
});

describe('#getPropertyName', () => {
	it('should find the correct node names', () => {
		expect(getInterfaceProperties(example1).map((node) => getPropertyName(node))).toEqual([]);

		expect(getInterfaceProperties(example2).map((node) => getPropertyName(node))).toEqual([
			'requestid',
			'ordereddependencies',
			'dependencyurls',
			'javascript',
			'paramvalues'
		]);

		expect(getInterfaceProperties(example3).map((node) => getPropertyName(node))).toEqual([
			'allowsort',
			'fieldname',
			'fieldtype',
			'indexanalyzer',
			'searchanalyzer',
			'similarity'
		]);

		expect(getInterfaceProperties(example4).map((node) => getPropertyName(node))).toEqual([
			'minzoom',
			'maxzoom',
			'radius',
			'extent',
			'nodesize',
			'log',
			'reduce',
			'initial',
			'map'
		]);
	});

	expect(getInterfaceProperties(example5).map((node) => getPropertyName(node))).toEqual([
		"firstname",
		"lastname"
	]);

});

describe('#getPropertyQuestion', () => {
	it('should find the correct node question mark', () => {
		expect(getInterfaceProperties(example1).map((node) => getPropertyQuestion(node))).toEqual([]);

		expect(getInterfaceProperties(example2).map((node) => getPropertyQuestion(node))).toEqual([
			false,
			true,
			true,
			true,
			false
		]);

		expect(getInterfaceProperties(example3).map((node) => getPropertyQuestion(node))).toEqual([
			true,
			false,
			true,
			true,
			true,
			true
		]);

		expect(getInterfaceProperties(example4).map((node) => getPropertyQuestion(node))).toEqual([
			true,
			true,
			true,
			true,
			true,
			true,
			true,
			true,
			true
		]);
	});
});

describe('#getPropertyKind', () => {
	it('should find the correct node question mark', () => {
		expect(getInterfaceProperties(example1).map((node) => getPropertyKind(node))).toEqual([]);

		expect(getInterfaceProperties(example2).map((node) => getPropertyKind(node))).toEqual([
			['number', 'string[]', 'Pizza'],
			['string[]'],
			['{[id:string]:string}'],
			['string'],
			['any[]']
		]);

		expect(getInterfaceProperties(example3).map((node) => getPropertyKind(node))).toEqual([
			['boolean'],
			['string'],
			['Field.FieldTypeEnum'],
			['string'],
			['string'],
			['Field.SimilarityEnum']
		]);

		expect(getInterfaceProperties(example4).map((node) => getPropertyKind(node))).toEqual([
			['number'],
			['number'],
			['number'],
			['number'],
			['number'],
			['boolean'],
			['Function'],
			['Function'],
			['Function']
		]);
	});
});

describe('#parseInterface', () => {
	it('should parse the interfaces correctly', () => {
		[example1, example2, example3, example4].forEach((example, idx) => {
			expect(parseInterface(example)).toMatchSnapshot(`example${idx + 1}`);
		});
	});
});
