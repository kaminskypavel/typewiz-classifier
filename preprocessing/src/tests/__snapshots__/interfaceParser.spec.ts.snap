// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#parseInterface should parse the interfaces correctly: example1 1`] = `
Object {
  "name": "IExample1",
  "structure": Array [],
}
`;

exports[`#parseInterface should parse the interfaces correctly: example2 1`] = `
Object {
  "name": "IExample2",
  "structure": Array [
    Object {
      "kind": Array [
        "number",
        "string[]",
        "Pizza",
      ],
      "name": "request_id",
      "question": false,
      "stem": "request_id",
    },
    Object {
      "kind": Array [
        "string[]",
      ],
      "name": "ordered_dependencies",
      "question": true,
      "stem": "ordered_depend",
    },
    Object {
      "kind": Array [
        "{[id:string]:string}",
      ],
      "name": "dependency_urls",
      "question": true,
      "stem": "dependency_url",
    },
    Object {
      "kind": Array [
        "string",
      ],
      "name": "java_script",
      "question": true,
      "stem": "java_script",
    },
    Object {
      "kind": Array [
        "any[]",
      ],
      "name": "param_values",
      "question": false,
      "stem": "param_valu",
    },
  ],
}
`;

exports[`#parseInterface should parse the interfaces correctly: example3 1`] = `
Object {
  "name": "IExample3",
  "structure": Array [
    Object {
      "kind": Array [
        "boolean",
      ],
      "name": "allow_sort",
      "question": true,
      "stem": "allow_sort",
    },
    Object {
      "kind": Array [
        "string",
      ],
      "name": "field_name",
      "question": false,
      "stem": "field_nam",
    },
    Object {
      "kind": Array [
        "Field.FieldTypeEnum",
      ],
      "name": "field_type",
      "question": true,
      "stem": "field_typ",
    },
    Object {
      "kind": Array [
        "string",
      ],
      "name": "index_analyzer",
      "question": true,
      "stem": "index_analyz",
    },
    Object {
      "kind": Array [
        "string",
      ],
      "name": "search_analyzer",
      "question": true,
      "stem": "search_analyz",
    },
    Object {
      "kind": Array [
        "Field.SimilarityEnum",
      ],
      "name": "similarity",
      "question": true,
      "stem": "similar",
    },
  ],
}
`;

exports[`#parseInterface should parse the interfaces correctly: example4 1`] = `
Object {
  "name": "IExample4",
  "structure": Array [
    Object {
      "kind": Array [
        "number",
      ],
      "name": "min_zoom",
      "question": true,
      "stem": "min_zoom",
    },
    Object {
      "kind": Array [
        "number",
      ],
      "name": "max_zoom",
      "question": true,
      "stem": "max_zoom",
    },
    Object {
      "kind": Array [
        "number",
      ],
      "name": "radius",
      "question": true,
      "stem": "radiu",
    },
    Object {
      "kind": Array [
        "number",
      ],
      "name": "extent",
      "question": true,
      "stem": "extent",
    },
    Object {
      "kind": Array [
        "number",
      ],
      "name": "node_size",
      "question": true,
      "stem": "node_s",
    },
    Object {
      "kind": Array [
        "boolean",
      ],
      "name": "log",
      "question": true,
      "stem": "log",
    },
    Object {
      "kind": Array [
        "Function",
      ],
      "name": "reduce",
      "question": true,
      "stem": "reduc",
    },
    Object {
      "kind": Array [
        "Function",
      ],
      "name": "initial",
      "question": true,
      "stem": "initi",
    },
    Object {
      "kind": Array [
        "Function",
      ],
      "name": "map",
      "question": true,
      "stem": "map",
    },
  ],
}
`;
